### PR TITLE
chore(ci): Upgrade dependencies and change package managers

### PR DESCRIPTION
**Pull request by peturgq**
####  To start off we're going to upgrade a few things:
- Upgrade Yarn to berry(2+)
- Upgrade Node to 18
- Upgrade Typescript to 4.4.2
- Upgrade graphql to 16.6.0
- Upgrade esbuild to 0.15.7
- Upgrade jest to 28.1.3
- Upgrade apollo-server-express to 3.10.2

Positive side effect of upgrading node and typescript is you are forced to update some other packages. 

Yarn is being upgraded to berry because this allows us to utilize some new methods to optimize things surrounding node_modules in our builds.

Typescript is being upgraded because of: https://github.com/yarnpkg/berry/issues/3275 
The changes introduced some more type safety enforcing in the compiler, so some additional changes in code were necessary.

## Issues pending fixes
- Running code coverage finds broken code which is usually caused by the new typescript compiler being stricter
## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Formatting passes locally with my changes
- [ ] I have rebased against main before asking for a review

### CODE REVIEW

**Review by disaerna — APPROVED** [review]

File: .gitignore [hunk 1]
```diff
@@ -15,6 +15,12 @@ libs/contentful-extensions/parent-slug/.cache
 
 # dependencies
 node_modules/
+.yarn/*
+!.yarn/patches
+!.yarn/releases
+!.yarn/plugins
+!.yarn/sdks
+!.yarn/versions
```
Comment (eirikurn) [inline .gitignore:23]: Why are we committing yarn releases and more in the monorepo?

File: apps/air-discount-scheme/web-e2e/src/integration/app.spec.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-describe('air-discount-scheme-web', () => {
+xdescribe('air-discount-scheme-web', () => {
```
Comment (eirikurn) [inline apps/air-discount-scheme/web-e2e/src/integration/app.spec.ts:1]: Whaaat? Remove instead?

File: apps/air-discount-scheme/web/i18n/withLocale.tsx [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import React from 'react'
 import { NextComponentType } from 'next'
-import { BaseContext, NextPageContext } from 'next/dist/next-server/lib/utils'
+import { BaseContext, NextPageContext } from 'next/dist/shared/lib/utils'
```
Comment (eirikurn) [inline apps/air-discount-scheme/web/i18n/withLocale.tsx:3]: Why are these changes needed if we didn't change the nextjs dependency?

File: apps/air-discount-scheme/web/tsconfig.json [hunk 4]
```diff
@@ -7,7 +7,8 @@
     "forceConsistentCasingInFileNames": true,
     "noEmit": true,
     "resolveJsonModule": true,
-    "isolatedModules": true
+    "isolatedModules": true,
+    "incremental": true
```
Comment (eirikurn) [inline apps/air-discount-scheme/web/tsconfig.json:11]: Exciting :)

File: apps/api/src/app/graphql.middleware.ts [hunk 5]
```diff
@@ -1,4 +1,5 @@
 import { FieldMiddleware, NextFn, MiddlewareContext } from '@nestjs/graphql'
+import { any } from 'zod'
```
Comment (eirikurn) [inline apps/api/src/app/graphql.middleware.ts:2]: Accidental?

File: apps/github-actions-cache/src/app/health.ts [hunk 6]
```diff
@@ -6,7 +6,7 @@ export const healthCheck = async (_: any, res: express.Response) => {
 }
 
 export const livenessCheck = async (_: any, res: express.Response) => {
-  if (cache.ping()) {
+  if (await cache.ping()) {
```
Comment (eirikurn) [inline apps/github-actions-cache/src/app/health.ts:9]: Yay typescript!

File: apps/judicial-system/backend/src/app/modules/case/case.service.ts [hunk 7]
```diff
@@ -993,7 +993,10 @@ export class CaseService {
   }
 
   async addCompletedCaseToQueue(caseId: string): Promise<string> {
-    return this.queueService.add({ type: MessageType.CASE_COMPLETED, caseId })
+    return await this.queueService.add({
+      type: MessageType.CASE_COMPLETED,
+      caseId,
+    })
```
Comment (eirikurn) [inline apps/judicial-system/backend/src/app/modules/case/case.service.ts:999]: Neccessary? Some recommend skipping await when you can return a "compatible" promise.

File: apps/judicial-system/web/src/routes/Prosecutor/components/PoliceCaseFiles/PoliceCaseFiles.tsx [hunk 8]
```diff
@@ -128,6 +128,7 @@ const PoliceCaseFiles: React.FC<Props> = ({
 
   const [policeCaseFiles, setPoliceCaseFiles] = useState<PoliceCaseFilesData>()
   useEffect(() => {
+    //if (typeof(policeDataError?.graphQLErrors[0].extensions?.code) !== "unknown")
```
Comment (eirikurn) [inline apps/judicial-system/web/src/routes/Prosecutor/components/PoliceCaseFiles/PoliceCaseFiles.tsx:131]: Remove?

File: apps/judicial-system/web/src/routes/Prosecutor/components/PoliceCaseFiles/PoliceCaseFiles.tsx [hunk 9]
```diff
@@ -151,7 +152,7 @@ const PoliceCaseFiles: React.FC<Props> = ({
         files: policeData ? policeData.policeCaseFiles : [],
         isLoading: false,
         hasError: true,
-        errorCode: policeDataError?.graphQLErrors[0].extensions?.code,
+        errorCode: policeDataError?.graphQLErrors[0].extensions.code as string,
```
Comment (eirikurn) [inline apps/judicial-system/web/src/routes/Prosecutor/components/PoliceCaseFiles/PoliceCaseFiles.tsx:155]: Possible regression. Is `extensions` always an object?

File: libs/clients/middlewares/src/lib/e2e/enhancedFetch.spec.ts [hunk 10]
```diff
@@ -1,3 +1,4 @@
+import 'node-fetch'
```
Comment (eirikurn) [inline libs/clients/middlewares/src/lib/e2e/enhancedFetch.spec.ts:1]: Why remove test-cert.pfx? The e2e test depends on it.

File: libs/infra-nest-server/src/lib/buildOpenApi.ts [hunk 11]
```diff
@@ -33,7 +33,7 @@ export const buildOpenApi = async ({
     // https://github.com/configcat/common-js/issues/36
     // TODO: Remove this when it's been fixed.
     process.exit(0)
-  } catch (e) {
+  } catch (e: any) {
```
Comment (eirikurn) [inline libs/infra-nest-server/src/lib/buildOpenApi.ts:36]: Should use `Error` instead?

File: libs/infra-next-server/src/lib/bootstrap.ts [hunk 12]
```diff
@@ -30,7 +30,7 @@ type BootstrapOptions = {
   /**
    * Proxy configuration. Ignored in production (according to NODE_ENV).
    */
-  proxyConfig?: { [context: string]: Config }
+  proxyConfig?: { [context: string]: any }
```
Comment (eirikurn) [inline libs/infra-next-server/src/lib/bootstrap.ts:33]: Doesn't the new version of http-proxy-middleware return a new type for this?

File: libs/infra-next-server/src/lib/bootstrap.ts [hunk 13]
```diff
@@ -84,7 +84,7 @@ export const bootstrap = async (options: BootstrapOptions) => {
     readyPromise,
     options.externalEndpointDependencies,
   )
-  expressApp.all('*', (req, res) => handle(req, res))
+  expressApp.all('*', (req: any, res: any) => handle(req, res))
```
Comment (eirikurn) [inline libs/infra-next-server/src/lib/bootstrap.ts:87]: We should have better express types 😬

File: libs/infra-next-server/src/lib/proxy.ts [hunk 14]
```diff
@@ -1,17 +1,13 @@
...
-  proxyConfig: { [context: string]: Config } | undefined,
+  proxyConfig: { [context: string]: any } | undefined,
   dev: boolean,
 ) => {
   if (!proxyConfig || !dev) {
     return
   }
-  const { default: createProxyMiddleware } = await import(
-    'http-proxy-middleware'
-  )
```
Comment (eirikurn) [inline libs/infra-next-server/src/lib/proxy.ts:14]: The idea here was to code-split this in production builds to not evaluate unused code on boot.

File: libs/shared/babel/src/exportFinder/exportFinder.ts [hunk 15]
```diff
@@ -133,7 +133,7 @@ function getExportDeclarations(
 }
 
 const getExports = (modulePath: string): ExportDefinition[] => {
-  const ast = parse(readFileSync(modulePath, 'utf-8'), {
+  const ast: any = parse(readFileSync(modulePath, 'utf-8'), {
```
Comment (eirikurn) [inline libs/shared/babel/src/exportFinder/exportFinder.ts:136]: Should not be needed. Might be an old babel duplication issue.

File: libs/shared/mocking/src/graphql/createResolvers.ts [hunk 16]
```diff
@@ -13,6 +13,7 @@ export type ResolverMap = Record<
   | Record<string, ((...args: any) => any) | undefined>
   | GraphQLScalarType
   | undefined
+  | any
```
Comment (eirikurn) [inline libs/shared/mocking/src/graphql/createResolvers.ts:16]: I don't like that we're adding `any` types across the codebase. I understand why, but maybe we can take a minute and find better fixes for some of them.

File: package.json [hunk 17]
```diff
@@ -4,8 +4,8 @@
   "license": "MIT",
   "engines": {
     "npm": "please-use-yarn",
-    "node": "^14.17",
-    "yarn": "^1.22"
+    "node": "18.8.0",
+    "yarn": "3.2"
```
Comment (eirikurn) [inline package.json:8]: 🙌🎉

File: package.json [hunk 18]
```diff
@@ -185,10 +186,10 @@
     "is-url": "1.2.4",
     "isomorphic-fetch": "3.0.0",
     "isomorphic-unfetch": "3.0.0",
+    "jest-environment-jsdom": "28.1.3",
```
Comment (eirikurn) [inline package.json:189]: devDependencies?

File: package.json [hunk 19]
```diff
@@ -268,41 +270,40 @@
...
-    "@graphql-codegen/typescript-operations": "1.17.15",
-    "@graphql-codegen/typescript-react-apollo": "1.17.7",
-    "@graphql-codegen/typescript-resolvers": "1.17.9",
+    "@graphql-codegen/cli": "2.12.0",
+    "@graphql-codegen/fragment-matcher": "3.3.1",
+    "@graphql-codegen/schema-ast": "2.5.1",
+    "@graphql-codegen/typescript": "2.7.3",
+    "@graphql-codegen/typescript-operations": "2.5.3",
+    "@graphql-codegen/typescript-react-apollo": "3.3.3",
+    "@graphql-codegen/typescript-resolvers": "2.7.3",
```
Comment (eirikurn) [inline package.json:306]: These are used by our schema generation script. No regressions? 🙈

Thankfully TypeScript should catch most of it.

File: apps/api/src/app/app.module.ts [hunk 20]
```diff
@@ -87,14 +89,17 @@ const autoSchemaFile = environment.production
...
       },
       plugins: [
-        responseCachePlugin({
-          shouldReadFromCache: ({ request: { http } }) => {
-            const bypassCacheKey = http?.headers.get('bypass-cache-key')
-            return bypassCacheKey !== process.env.BYPASS_CACHE_KEY
-          },
-        }),
+        // This was causing problems since graphql upgrade, gives us issues like:
+        // Error: overallCachePolicy.policyIfCacheable is not a function
```
Comment (eirikurn) [inline apps/api/src/app/app.module.ts:93]: Sounds like a node_modules resolution/deduplication issue. 

Not sure how important this cache is but would feel better if we fix it :)

**Review by eirikurn — COMMENTED** [review]
Great job!

File: apps/auth-admin-web/specs/index.spec.tsx [hunk 21]
```diff
@@ -1,6 +1,6 @@
 import React from 'react'
 import { render } from '@testing-library/react'
-
+import 'whatwg-fetch'
```
Comment (oddsson) [inline apps/auth-admin-web/specs/index.spec.tsx:3]: Is this accidental?

File: apps/judicial-system/web/src/routes/Prosecutor/components/PoliceCaseFiles/PoliceCaseFiles.tsx (hunk 8 above)
Comment (oddsson) [inline apps/judicial-system/web/src/routes/Prosecutor/components/PoliceCaseFiles/PoliceCaseFiles.tsx:131]:
```suggestion
```

File: apps/judicial-system/web/src/routes/Prosecutor/components/PoliceCaseFiles/PoliceCaseFiles.tsx (hunk 9 above)
Comment (oddsson) [inline apps/judicial-system/web/src/routes/Prosecutor/components/PoliceCaseFiles/PoliceCaseFiles.tsx:155]:
```suggestion
        errorCode: policeDataError?.graphQLErrors[0].extensions?.code as string,
```

Extensions is of type `readonly extensions: { [key: string]: any } | undefined;`. It can be `undefined` so I'd think we need a `?` after `extensions`

File: apps/judicial-system/backend/src/app/modules/case/case.service.ts (hunk 7 above)
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/case/case.service.ts:999]: Agree, there is no need to wait.

**Review by gudjong — CHANGES_REQUESTED** [review]
One thing

File: apps/judicial-system/web/src/routes/Prosecutor/components/PoliceCaseFiles/PoliceCaseFiles.tsx (hunk 9 above)
Comment (Arnarkari93) [inline apps/judicial-system/web/src/routes/Prosecutor/components/PoliceCaseFiles/PoliceCaseFiles.tsx:155]: I think adding a few `?.` should do the trick here. `extensions` if defined is always an object/map https://spec.graphql.org/draft/#sel-FAPHRPDZAADbBAzwM
```suggestion
        errorCode: policeDataError?.graphQLErrors[0]?.extensions?.code as string,
```

File: apps/judicial-system/web/src/routes/Shared/SignedVerdictOverview/Components/ModifyDatesModal/ModifyDatesModal.spec.ts [hunk 22]
```diff
@@ -5,7 +5,7 @@ import { createCaseModifiedExplanation } from './ModifyDatesModal'
 describe('createCaseModifiedExplanation', () => {
   const formatMessage = createIntl({ locale: 'is', onError: jest.fn() })
     .formatMessage
-  beforeAll(() => jest.useFakeTimers('modern'))
+  beforeAll(() => jest.useFakeTimers())
```
Comment (Arnarkari93) [inline apps/judicial-system/web/src/routes/Shared/SignedVerdictOverview/Components/ModifyDatesModal/ModifyDatesModal.spec.ts:8]: It seems like the new way of setting the time in jest tests is setting the `now` field in the `fakeTimersConfig` 
```suggestion
  beforeAll(() => jest.useFakeTimers({ now: new Date('2022-06-13T13:37:00Z') }))
```
https://jestjs.io/docs/jest-object#fake-timers

File: apps/judicial-system/web/src/utils/validate.spec.ts [hunk 23]
```diff
@@ -185,7 +185,7 @@ describe('Validate phonenumber format', () => {
 
 describe('Validate court case number', () => {
   beforeAll(() => {
-    jest.useFakeTimers('modern').setSystemTime(new Date('2020-01-01'))
+    jest.useFakeTimers().setSystemTime(new Date('2020-01-01'))
```
Comment (Arnarkari93) [inline apps/judicial-system/web/src/utils/validate.spec.ts:188]:
```suggestion
    jest.useFakeTimers({ now: new Date('2020-01-01') })
```

**Review by Arnarkari93 — CHANGES_REQUESTED** [review]

**Review by obmagnusson — APPROVED** [review]

File: .gitignore (hunk 1 above)
Comment (peturgq) [reply .gitignore:23]: `yarn/releases` contains the executable for yarn, the other ones I don't know so much, but I took this from here:
https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored

File: apps/api/src/app/graphql.middleware.ts (hunk 5 above)
Comment (peturgq) [reply apps/api/src/app/graphql.middleware.ts:2]: Absolutely accidental, good catch!

File: package.json (hunk 19 above)
Comment (peturgq) [reply package.json:306]: We are generating schemas successfully but I'm not sure how to test them specifically beyond that, would love your input in that area.

File: apps/auth-admin-web/specs/index.spec.tsx (hunk 21 above)
Comment (peturgq) [reply apps/auth-admin-web/specs/index.spec.tsx:3]: No, seemed like the only way to get fetching to work inside the test, on fetch the test is using relative paths where the default global fetch API introduced in v18 was failing, I tried `node-fetch` too but this was also failing.

**Comment by peturgq** [comment]
@gudjong 
> One thing 
> Agree, there is no need to wait.

If we did not use await there we hit the following error:
'Can not use queue until nest application bootstrapping has finished'

**Comment by gudjong** [comment]
> @gudjong
> 
> > One thing
> > Agree, there is no need to wait.
> 
> If we did not use await there we hit the following error: 'Can not use queue until nest application bootstrapping has finished'

That is surprising, but I will accept the change. This code has change and we may need to revisit this when you merge with main.

**Review by gudjong — APPROVED** [review]

File: apps/air-discount-scheme/web/i18n/withLocale.tsx (hunk 3 above)
Comment (peturgq) [reply apps/air-discount-scheme/web/i18n/withLocale.tsx:3]: I don't think this ever worked, its broken on main f.ex, `lib/utils` simply does not exist in `next-server` in this version of next.
I just took the liberty of fixing it here.

File: apps/judicial-system/web/src/routes/Shared/SignedVerdictOverview/Components/ModifyDatesModal/ModifyDatesModal.spec.ts (hunk 22 above)
Comment (peturgq) [reply apps/judicial-system/web/src/routes/Shared/SignedVerdictOverview/Components/ModifyDatesModal/ModifyDatesModal.spec.ts:8]: Are we ok with using something arbitrary like '2022-06-13T13:37:00Z' here?

Comment (Arnarkari93) [reply apps/judicial-system/web/src/routes/Shared/SignedVerdictOverview/Components/ModifyDatesModal/ModifyDatesModal.spec.ts:8]: yes, for this test it is fine.

Comment (peturgq) [reply apps/judicial-system/web/src/routes/Shared/SignedVerdictOverview/Components/ModifyDatesModal/ModifyDatesModal.spec.ts:8]: Seeing as these tests are succeeding, I think we should let it slip for now, this PR is not supposed to go into changing the functionality of tests.

My assumption was that the faketimers now use 'modern' by default, I will let this one be while I fixed the other one where a date was being used.

File: apps/api/src/app/app.module.ts (hunk 20 above)
Comment (peturgq) [reply apps/api/src/app/app.module.ts:93]: We tried fixing this, but weren't successful - Any ideas on what this cache actually does?

**Review by Arnarkari93 — APPROVED** [review]

**Comment by peturgq** [comment]
We're already seeing failures now after merging latest main, here.. We'll fix this and merge this in ASAP
